server-common/src/main/java/org/apache/gravitino/server/authorization/MetadataFilterHelper.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.gravitino.server.authorization;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.gravitino.MetadataObject;
+import org.apache.gravitino.server.authorization.expression.AuthorizationExpressionExecutor;
+
+/**
+ * MetadataFilterHelper performs permission checks on the list data returned by the REST API based
+ * on expressions or metadata types, and calls {@link GravitinoAuthorizer} for authorization,
+ * returning only the metadata that the user has permission to access.
+ */
+public class MetadataFilterHelper {
+
+  private MetadataFilterHelper() {}
+
+  /**
+   * Call {@link GravitinoAuthorizer} to filter the metadata list
+   *
+   * @param metadataType for example, CATALOG, SCHEMA,TABLE, etc.
+   * @param privilege for example, CREATE_CATALOG, CREATE_TABLE, etc.
+   * @param metadataList metadata list.
+   * @param metadataIdExtractor extract the metadata id from list items.
+   * @return metadata List that the user has permission to access.
+   * @param <E> metadata object.
+   */
+  public static <E> List<E> filter(
+      MetadataObject.Type metadataType,
+      String privilege,
+      List<E> metadataList,
```

Comment:
What's the relation between E and MetataObject.Type? Is it redundant?

Reply:
It is used for the REST API that performs listing, where E represents the different types of metadata entities that the REST API is originally intended to return.

Reply:
Should E extends MetadataObject?

Reply:
If we have already had a metadataList, we can get every metadataObject type, why do we need a extra paramter `metadataType`?

Reply:
> Should E extends MetadataObject?
Currently, the results of the list operation do not extend from **MetadataObject**.
<img width="869" alt="image" src="https://github.com/user-attachments/assets/0329faa2-3beb-4a19-882e-e42c4723e26a" />
<img width="783" alt="image" src="https://github.com/user-attachments/assets/f73ac5e1-b637-457b-be4f-ce84011940c4" />

Reply:
So we should have two methods NameIdentifier and MetadataObjects?

Reply:
I think we only need return `NameIdentifier`, Because metadata authz only process RESTful interface and `MetadataObjects` only use in internal?
@jerqi What do you think?

Reply:
> I think we only need return `NameIdentifier`, Because metadata authz only process RESTful interface and `MetadataObjects` only use in internal? @jerqi What do you think?

The code below returns **Catalog[]** instead of **NameIdentifier[]**.
<img width="1772" alt="image" src="https://github.com/user-attachments/assets/68dbb2d0-7c03-4629-9098-72db87f4d1c3" />